internal/resources/iampolicy/scope.go

```diff
@@ -0,0 +1,197 @@
+)
+
+var (
+	scopeSchema = &schema.Schema{
+		Type:        schema.TypeList,
+		Description: "Scope of the resource on which the rolebinding has to be added, having one of the valid scopes: organization, cluster_group, cluster, workspace or namespace.",
+		Required:    true,
+		ForceNew:    true,
+		MinItems:    1,
+		MaxItems:    1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				organizationKey: organizationFullname,
+				clusterGroupKey: clusterGroupFullname,
+				clusterKey:      clusterFullname,
+				workspaceKey:    workspaceFullname,
+				namespaceKey:    namespaceFullname,
+			},
+		},
+	}
+	scopesAllowed = [...]string{organizationKey, clusterGroupKey, clusterKey, workspaceKey, namespaceKey}
+)
+
+type (
+	scope int64
+	// ScopedFullname is a struct for all types of policy full names.
+	scopedFullname struct {
+		scope                scope
+		fullnameOrganization *organizationmodel.VmwareTanzuManageV1alpha1OrganizationFullName
+		fullnameClusterGroup *clustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupFullName
+		fullnameCluster      *clustermodel.VmwareTanzuManageV1alpha1ClusterFullName
+		fullnameWorkspace    *workspacemodel.VmwareTanzuManageV1alpha1WorkspaceFullName
+		fullnameNamespace    *namespacemodel.VmwareTanzuManageV1alpha1ClusterNamespaceFullName
+	}
+)
+
+func constructScope(d *schema.ResourceData) (scopedFullnameData *scopedFullname) {
+	value, ok := d.GetOk(scopeKey)
+	if !ok {
+		return scopedFullnameData
+	}
+
+	data, _ := value.([]interface{})
+
+	if len(data) == 0 || data[0] == nil {
+		return scopedFullnameData
+	}
+
+	scopeData := data[0].(map[string]interface{})
+
+	if v, ok := scopeData[organizationKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:                organizationScope,
+				fullnameOrganization: constructOrganizationFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[clusterGroupKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:                clusterGroupScope,
+				fullnameClusterGroup: constructClusterGroupFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[clusterKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:           clusterScope,
+				fullnameCluster: constructClusterFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[workspaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:             workspaceScope,
+				fullnameWorkspace: constructWorkspaceFullname(v1),
+			}
+		}
+	}
+
+	if v, ok := scopeData[namespaceKey]; ok {
+		if v1, ok := v.([]interface{}); ok && len(v1) != 0 {
+			scopedFullnameData = &scopedFullname{
+				scope:             namespaceScope,
+				fullnameNamespace: constructNamespaceFullname(v1),
+			}
+		}
+	}
```

Comment:
Please import and use the construct full name functions as the base resources are already implemented.